ext/MitsubaLoader/CMitsubaMaterialCompilerFrontend.cpp

```diff
@@ -0,0 +1,597 @@
+
+            getSpectrumOrTexture(current->difftrans.transmittance, node->transmittance);
+        }
+        break;
+        case CElementBSDF::PLASTIC:
+        case CElementBSDF::ROUGHPLASTIC:
+        {
+            nextSym = ir->allocNode<IR::CBSDFCombinerNode>(IR::CBSDFCombinerNode::ET_DIFFUSE_AND_SPECULAR);
+            nextSym->children.count = 2u;
+
+            const float eta = current->plastic.intIOR/current->plastic.extIOR;
+
+            auto& diffuse = nextSym->children[0];
+            diffuse = ir->allocNode<IR::CMicrofacetDiffuseBSDFNode>();
+            auto& specular = nextSym->children[1];
+            specular = ir->allocNode<IR::CMicrofacetSpecularBSDFNode>();
+
+            auto* node_specular = static_cast<IR::CMicrofacetSpecularBSDFNode*>(specular);
+            node_specular->shadowing = IR::CMicrofacetSpecularBSDFNode::EST_SMITH;
+            node_specular->eta = IR::INode::color_t(eta);
+            node_specular->scatteringMode = IR::CMicrofacetSpecularBSDFNode::ESM_REFLECT;
+            if (currType == CElementBSDF::ROUGHPLASTIC)
+            {
+                node_specular->ndf = ndfMap[current->plastic.distribution];
+                getFloatOrTexture(current->plastic.alphaU, node_specular->alpha_u);
+                if (node_specular->ndf == IR::CMicrofacetSpecularBSDFNode::ENDF_ASHIKHMIN_SHIRLEY)
+                    getFloatOrTexture(current->plastic.alphaV, node_specular->alpha_v);
+                else
+                    node_specular->alpha_v = node_specular->alpha_u;
+            }
+            else
+            {
+                node_specular->setSmooth();
+            }
+            auto* node_diffuse = static_cast<IR::CMicrofacetDiffuseBSDFNode*>(diffuse);
+            getSpectrumOrTexture(current->plastic.diffuseReflectance, node_diffuse->reflectance);
+            node_diffuse->alpha_u = node_specular->alpha_u;
+            node_diffuse->alpha_v = node_specular->alpha_v;
+            node_diffuse->eta = node_specular->eta;
+        }
+        break;
+        case CElementBSDF::DIELECTRIC:
+        case CElementBSDF::THINDIELECTRIC:
+        case CElementBSDF::ROUGHDIELECTRIC:
+        {
+            nextSym = ir->allocNode<IR::CBSDFCombinerNode>(IR::CBSDFCombinerNode::ET_FRESNEL_BLEND);
+            nextSym->children.count = 2u;
+
+            const float eta = current->dielectric.intIOR/current->dielectric.extIOR;
+            _IRR_DEBUG_BREAK_IF(eta==1.f);
+            if (eta==1.f)
+                os::Printer::log("WARNING: Dielectric with IoR=1.0!", current->id, ELL_ERROR);
+
+            auto& refl = nextSym->children[0];
+            refl = ir->allocNode<IR::CMicrofacetSpecularBSDFNode>();
+            auto& trans = nextSym->children[1];
+            trans = ir->allocNode<IR::CMicrofacetSpecularBSDFNode>();
+
+            auto* node_refl = static_cast<IR::CMicrofacetSpecularBSDFNode*>(refl);
+            auto* node_trans = static_cast<IR::CMicrofacetSpecularBSDFNode*>(trans);
```

Comment:
explain the design decision behind having two separate nodes for what is essentially a single BSDF?

Reply:
flexibility of IR, compatibility with MDL. I mean it could have Dielectric node, but scattering mode is also useful (and needed to make it work with MDL) so since we have scattering mode i decided to have dielectric be encoded like this

Reply:
so how do you see that you need to match front and back into one thing?

Reply:
its not front and back, its reflection and transmission

Reply:
can you show me an example of an MDL definition of a material that would require this?

Reply:
I also take issue with the fact that direct childre of this `Fresnel` combiner node are `CMicrofacetSpecularBSDFNode` which when used directly are supposed to handle their own fresnel?

Reply:
?